File: week2/algorithm_sort/merge_sort/merge.c
#include <stdio.h>
#include <math.h>

void merge_sort(int arr[], int l);
void merge(int result[], int left[], int l_length, int right[], int r_length);

int main(void)
{
    int arr[] = {5, 2, 1, 3, 6, 4};
    const int L = sizeof(arr) / sizeof(arr[0]);

    merge_sort(arr, L);

    printf("Sorted Array: ");
    for (int i = 0; i < L; i++)
    {
        printf("  %i", arr[i]);
    }
    printf("\n");
}

void merge_sort(int arr[], int L)
{
    if (L > 1) {
        int middle = floor(L / 2);
        int left[middle];
        int right_length = L % 2 ? middle + 1 : middle;
        int right[right_length];

        for (int i = 0; i < middle; i++) {
            left[i] = arr[i];
        }

        int index = 0;
        for (int i = middle; i < L; i++) {
            right[index] = arr[i];
            index++;
        }


        merge_sort(left, middle);
        merge_sort(right, right_length);

        for (int i = 0; i < L; i++)
        {
            for (int j = i + 1; j < L; j++)
            {
                if (arr[i] > arr[j])
                {
                    arr[i] = arr[j];
                    arr[j] = arr[i];
                }
            }
        }


        merge(arr, left, middle, right, right_length);
    }
}

void merge(int result[], int left[], int l_length, int right[], int r_length)
{
    int left_index = 0;
    int right_index = 0;
    int index = 0;

    while (left_index < l_length && right_index < r_length)
    {
        if (left[left_index] > right[right_index])
        {
            result[index] = right[right_index];
            right_index++;
        } else {
            result[index] = left[left_index];
            left_index++;
        }

        index++;
    }

    while (left_index < l_length)
    {
        result[index] = left[left_index];
        left_index++;
        index++;
    }

    while (right_index < r_length)
    {
        result[index] = right[right_index];
        right_index++;
        index++;
    }
}

File: week2/algorithm_sort/merge_sort/merge.ts
function merge_sort(arr: number[]): number[] {
  const L = arr.length;

  if (L < 2) {
    return arr;
  }

  const LEFT_L = Math.floor(L/2);
  const RIGHT_L = L % 2 ? LEFT_L + 1 : LEFT_L;
  let arr_sorted = [];
  let left = [];
  let right = [];

  for (let j = 0; j < RIGHT_L; j++) {
    if (j < LEFT_L) {
      left[j] = arr[j];
    }

    if (j < RIGHT_L) {
      right[j] = arr[LEFT_L + j];
    }
  }

  for (let i = 0; i < LEFT_L - 1; i++) {
    for (let j = i + 1; j < LEFT_L; j++) {
      if (left[i] > left[j]) {
        [left[i], left[j]] = [left[j], left[i]];
      }
    }
  }

  for (let i = 0; i < RIGHT_L - 1; i++) {
    for (let j = i + 1; j < RIGHT_L; j++) {
      if (right[i] > right[j]) {
        [right[i], right[j]] = [right[j], right[i]];
      }
    }
  }

  for (let i = 0; i < LEFT_L; i++) {
    for (let j = 0; j < RIGHT_L; j++) {
      if (!right[j]) {
        continue;
      }

      if (left[i] > right[j]) {
        arr_sorted.push(right[j]);
        right[j] = null;
      } else {
        arr_sorted.push(left[i]);
        left[i] = null;

        if (i === LEFT_L - 1) {
            arr_sorted.push(right[j]);
            right[j] = null;
        }

        break;
      }
    }

    if (left[i]) {
      arr_sorted.push(left[i]);
    }
  }

  return arr_sorted;
}

console.log(merge_sort([5]));
console.log(merge_sort([5, 2]));
console.log(merge_sort([5, 2, 1]));
console.log(merge_sort([5, 2, 1, 3]));
console.log(merge_sort([5, 2, 1, 3, 6]));
console.log(merge_sort([5, 2, 1, 3, 6, 4]));
